sintr_analysis_server_example/lib/completion_metrics.dart

```diff
@@ -0,0 +1,228 @@
+/// and removed when a matching reponse is found
+Map<String, _Completion> _requestMap = <String, _Completion>{};
+
+String _composeExtractionResult(_Completion completion, int completionTime) =>
+    '$extractedSdkVersion,$completionTime';
+
+/// Search the given [logMessageText] for the given [key]
+/// and return the associated value.
+String _extractJsonValue(String logMessageText, String key) {
+  var prefix = '"$key"::';
+  int start = logMessageText.indexOf(prefix);
+  if (start == -1) throw 'expected $key in $logMessageText';
+  start += prefix.length;
+  int deliminator = logMessageText.codeUnitAt(start);
+  if (deliminator == QUOTE) {
+    // Return quoted string
+    ++start;
+    int end = logMessageText.indexOf('"', start);
+    if (end == -1) throw 'expected value for $key in $logMessageText';
+    return logMessageText.substring(start, end);
+  } else if (deliminator == OPEN_BRACE) {
+    // Return JSON map
+    // TODO (danrubel) handle nested and embedded braces
+    int end = logMessageText.indexOf('}');
+    if (end == -1) throw 'expected matching brace for $key in $logMessageText';
+    return logMessageText.substring(start, end + 1);
+  }
+  return null;
+}
+
+/// Process a log message and return a string representing the result
+/// or `null` if no result from the given log entry.
+String _processLogMessage(int time, String msgType, String logMessageText) {
+  if (msgType == 'Req') {
+    String method = _extractJsonValue(logMessageText, 'method');
+    return _processRequest(time, method, logMessageText);
+  } else if (msgType == 'Res') {
+    String requestId = _extractJsonValue(logMessageText, 'id');
+    return _processResponse(time, requestId, logMessageText);
+  } else if (msgType == 'Noti') {
+    String event = _extractJsonValue(logMessageText, 'event');
+    return _processNotification(time, event, logMessageText);
+  } else if (msgType == 'Read') {
+    return null;
+  } else if (msgType == 'Task') {
+    return null;
+  } else if (msgType == 'Log') {
+    return null;
+  } else if (msgType == 'Perf') {
+    return null;
+  } else if (msgType == 'Watch') {
+    return null;
+  }
+  //throw 'unknown msgType $msgType in $logMessageText';
+  return null;
+}
+
+/// Process a log entry representing a notification
+String _processNotification(int time, String event, String logMessageText) {
+  if (event == 'completion.results') {
+    if (!logMessageText.endsWith(',"isLast"::true}}')) return null;
+    var prefix = '{"event"::"completion.results","params"::{"id"::"';
+    if (!logMessageText
+        .startsWith(prefix)) throw 'expected $prefix in $logMessageText';
+    int start = prefix.length;
+    int end = logMessageText.indexOf('"', start);
+    if (end == -1) throw 'expected notification id in $logMessageText';
+    String notificationId = logMessageText.substring(start, end);
+    _Completion completion = _notificationMap.remove(notificationId);
+    if (completion ==
```

Comment:
Silly nit, I don't think the auto formatter is doing a good job with these too long lines. Breaking inside the if clause seems really unfortunate for reading the intention.

Consider if (x) { 
  throw " .... " 
}

Instead to stop it doing this.